### PR TITLE
Task/1937

### PR DESCRIPTION
**Pull request by mustaphazorgati**
https://sonarcloud.io/summary/new_code?id=mustaphazorgati_taskana&branch=task%2F1937
---
Release Notes:

```
**NEW**
* TSK-1937: added AfterRequestReviewProvider which allows customized behavior after a review has been requested on a given Task. 

```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [x] PR fulfills the ticket
- [x] Edge cases and unwanted side effects are tested
- [x] Readability

### CODE REVIEW

**Review by gitgoodjhe — APPROVED** [review]

File: lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+  @Nested
+  @TestInstance(Lifecycle.PER_CLASS)
+  class ServiceProviderInstantiation {
+
+    @Test
+    void should_InstantiateServiceProvider_When_ServiceProviderIsTopLevelClass() {
+      @WithServiceProvider(
+          serviceProviderInterface = CreateTaskPreprocessor.class,
+          serviceProviders = TopLevelCreateTaskPreprocessor.class)
+      class Clazz {}
```
Comment (ryzheboka) [inline lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java:79]: The name "Clazz" sounds unclear. Maybe we could rename it to "ExampleTestClass", "ExampleTestClassUsingServiceProviders" or something else?

File: lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+      assertThatCode(call).doesNotThrowAnyException();
+    }
+
+    @Test
+    void should_InstantiateServiceProvider_When_ServiceProviderIsStaticMemberClass() {
+      @WithServiceProvider(
+          serviceProviderInterface = CreateTaskPreprocessor.class,
+          serviceProviders = StaticCreateTaskPreprocessor.class)
+      class Clazz {}
```
Comment (ryzheboka) [inline lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java:91]: What do you think about changing the class name?

File: lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+      assertThatCode(call).doesNotThrowAnyException();
+    }
+
+    @Test
+    void should_InstantiateServiceProvider_When_ServiceProviderIsPrivateStaticMemberClass() {
+      @WithServiceProvider(
+          serviceProviderInterface = CreateTaskPreprocessor.class,
+          serviceProviders = PrivateStaticCreateTaskPreprocessor.class)
+      class Clazz {}
```
Comment (ryzheboka) [inline lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java:103]: What do you think about changing the class name?

File: lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    @Test
+    void should_InstantiateServiceProvider_When_ServiceProviderIsTopLevelClass() {
+      @WithServiceProvider(
+          serviceProviderInterface = CreateTaskPreprocessor.class,
+          serviceProviders = TopLevelCreateTaskPreprocessor.class)
+      class Clazz {}
+
+      ThrowingCallable call = () -> extractServiceProviders(Clazz.class);
+
+      assertThatCode(call).doesNotThrowAnyException();
```
Comment (ryzheboka) [inline lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java:83]: Should we additionally check that the return value is not empty? Or that the return value is a list containing an instance of TopLevelCreateTaskPreprocessor?

Why do you check it in ExtractMultipleServiceProvidersFromMultipleServiceProviderInterfaces, but not here?

**Review by ryzheboka — CHANGES_REQUESTED** [review]

**Review by ryzheboka — APPROVED** [review]

**Pull request merged**